File: haskell/binary/Binary.hs
module Binary (toDecimal) where

toDecimal :: String -> Int
toDecimal digits = parseDigits digits 0

parseDigits :: String -> Int -> Int
parseDigits "" value = value
parseDigits (digit:remainder) value =
    value `seq`
    case digit of
		'0' -> parseDigits remainder (value * 2)
		'1' -> parseDigits remainder (value * 2 + 1)
		_   -> 0

File: vbnet/binary/Binary.vb
Public Class Binary
    Private ReadOnly _digits As Char()
    
    Public Sub New(binary As String)
        _digits = binary.ToCharArray()
    End Sub

    Public Function ToDecimal() As Integer
        Dim dec = 0
        
        For Each digit in _digits
            If digit = "0"
               dec = dec << 1
            ElseIf digit = "1"
                dec = (dec << 1) + 1
            Else 
                Return 0
            End If
        Next
        
        Return dec
    End Function
End Class
